homeassistant/components/broadlink/remote.py

```diff
@@ -0,0 +1,418 @@
+            raise
+
+        try:
+            self._commands[device][command] = code
+        except KeyError:  # New device
+            self._commands[device] = {command: code}
+            self._toggle[device] = 0
+        return True
+
+    async def _async_capture_code(
+        self, command, toggle=False, timeout=DEFAULT_LEARNING_TIMEOUT
+    ):
+        """Enter learning mode and capture a code from a remote.
+
+        Capture an aditional code for toggle commands.
+        """
+        await self._async_attempt(self._api.enter_learning)
+
+        self.hass.components.persistent_notification.async_create(
+            f"Press the '{command}' button.",
+            title="Learn command",
+            notification_id="learn_command",
+        )
+
+        code = None
+        start_time = utcnow()
+        while (utcnow() - start_time) < timedelta(seconds=timeout):
+            code = await self.hass.async_add_executor_job(self._api.check_data)
+            if code:
+                break
+            await asyncio.sleep(1)
+
+        self.hass.components.persistent_notification.async_dismiss(
+            notification_id="learn_command"
+        )
+
+        if not code:
+            raise TimeoutError
+        if all(not value for value in code):
+            raise ValueError
+
+        code = b64encode(code).decode("utf8")
+        return (
+            code
+            if not toggle
+            else [code, await self._async_capture_code(command, timeout=timeout)]
+        )
+
+    async def _async_attempt(self, function, *args):
+        """Retry a socket-related function until it succeeds."""
+        for retry in range(DEFAULT_RETRY):
+            if retry and not await self._async_connect():
+                continue
+            try:
+                await self.hass.async_add_executor_job(function, *args)
+            except socket.error:
+                continue
+            return
+        raise ConnectionError
+
+    async def _async_connect(self):
+        """Connect to the remote."""
+        try:
+            auth = await self.hass.async_add_executor_job(self._api.auth)
+        except socket.error:
+            auth = False
+        if auth and not self._available:
+            _LOGGER.warning("Connected to the remote")
+            self._available = True
+        elif not auth and self._available:
+            _LOGGER.warning("Disconnected from the remote")
+            self._available = False
+        return auth
+
+    async def async_retrieve_data(self):
+        """Load dictionary of commands and toggle data from files."""
+        try:
+            await self._async_load_commands(self._command_file)
+            await self._async_load_toggle_data(self._toggle_file)
+        except FileNotFoundError:
+            pass  # No problem. The file will be created soon.
+        except (IOError, json.JSONDecodeError, vol.MultipleInvalid):
```

Comment:
`IOError` is deprecated. Use `OSError` instead.

Reply:
I changed the storage method and now this exception is handled in homeassistant.util.json.load_json(), so I deleted this part of the code. But thanks for the advice.